frontends/p4/alias.h

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2019 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _FRONTENDS_P4_ALIAS_H_
+#define _FRONTENDS_P4_ALIAS_H_
+
+/*
+ * A simple conservative syntactic alias analysis.  Two things may
+ * alias if they refer to objects with the same name.  This pass may
+ * be run early in the front end, before enough information is present
+ * (eg.  def-use information) to do a precise alias analysis.  Also,
+ * this pass is safe only if the expressions compared for aliasing are
+ * part of the *same statement*.
+ */
+
+#include "ir/ir.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/methodInstance.h"
+
+namespace P4 {
+
+/// This class represents the path to a location.
+/// Given a struct S { bit a; bit b; } and a variable S x;
+/// a path can be x.a, or just x.  An array index is represented as a
+/// number (encoded as a string) or as "*", denoting an unknown index.
+struct LocationPath : public IHasDbPrint {
+    const IR::IDeclaration* root;
+    std::vector<cstring> path;
+
+    explicit LocationPath(const IR::IDeclaration* root): root(root) { CHECK_NULL(root); }
+
+    const LocationPath* append(cstring suffix) const {
+        auto result = new LocationPath(root);
+        result->path = path;
+        result->path.push_back(suffix);
+        return result;
+    }
+
+    /// True if this path is a prefix of other or the other way around
+    bool isPrefix(const LocationPath* other) const {
+        // Due to the structure of the P4 language, two distinct
+        // declarations can never alias.
+        if (root != other->root)
+            return false;
+        size_t len = std::min(path.size(), other->path.size());
+        for (size_t i = 0; i < len; i++) {
+            if (path.at(i) == "*" || other->path.at(i) == "*")
+                continue;
+            if (path.at(i) != other->path.at(i))
+                return false;
+        }
+        return true;
+    }
+
+    void dbprint(std::ostream& out) const override {
+        out << root->getName();
+        for (auto p : path)
+            out << "." << p;
+    }
+};
+
+/// We represent a set of location set as a set of LocationPath
+/// objects.
+class SetOfLocations : public IHasDbPrint {
+ public:
+    std::set<const LocationPath*> paths;
```

Comment:
nit: is there a need for the indirection here, or could this be `std::set<LocationPath>` instead?  Is a single LocationPath object ever required in two sets such that it being the same object instead of a copy important?

Reply:
No, we don't even test the LocationPaths for equality really; there may be probably some duplicates. This should not affect correctness. So we can probably use a vector instead of a set.